### DISCUSSION TITLE
Compiler giving "The compiler is unable to type-check this expression in reasonable time; try breaking up the expression into distinct sub-expressions" issue when using multiple sheets with  Destination Reducer approach

### DISCUSSION THREAD

**Discussion by parvBhaskarStryker** [pointfreeco/swift-composable-architecture#2922] [Q&A] [answered]
Hi team

I'm using Destination(4) reducer feature to combine all my presentation sheets, So in my view I have some normal sheets and some grouped Destination reducer sheet.
I'm getting issue when I'm tring to add 5th sheet in my Destination reducer. it gives me above error, when I remove or comment any existing destination sheet then the code compiles and run with the new sheet.
So **I want to know if we have some limit in the Destination reducer ?**

@Reducer
struct SampleFeature {
@Presents var destination: Destination.State?
@Presents var showRandomView : RandomViewFeature.State?
@Presents var showSettingsView: SettingsViewFeature.State?
var code: String = ""
}
enum Action: Equatable {
        case destination(PresentationAction<Destination.Action>)
        case showRandomView(PresentationAction<RandomViewFeature.Action>)
        case settingsView(PresentationAction<SettingsViewFeature.Action>)
    }
    @Reducer(state: .equatable, action: .equatable)
    enum Destination {
        case view1(view1Feature)
        case view2(view2Feature)
        case view3(view3Feature)
        case view4(view4Feature)
        case view5(view5Feature)
    }

 var body: some Reducer<State, Action> {
EmptyReducer()
}
.ifLet(\.$destination, action: \.destination)

@Bindable var store: Store<SampleFeature.State, SampleFeatureAction>
var body: some View {
      .....
}
.sheet(item: $store.scope(state: \.showRandomView, action: \.showRandomView)) { sheetStore in
            RandomViewFeature(store: sheetStore)
        }
        .sheet(
            item: $store.scope(state: \.destination?.view1, action: \.destination.view1)
        ) { sheetStore in
            View1(store: sheetStore)
                .presentationDetents([.medium])
        }
        .sheet(
            item: $store.scope(state: \.destination?.view2, action: \.destination.view2)
        ) { sheetStore in
            View2(store: sheetStore)
                .presentationDetents([.medium])
                .interactiveDismissDisabled()
        }
        .sheet(
            item: $store.scope(state: \.destination?.view3, action: \.destination.view3)
        ) { sheetStore in
            View3(store: sheetStore)
                .presentationDetents([.medium])
                .presentationDragIndicator(.hidden)
        }
        .sheet(
            item: $store.scope(state: \.destination?.view4, action: \.destination.view4)
        ) { sheetStore in
            View4(store: sheetStore)
                .presentationDetents([.medium])
                .presentationDragIndicator(.hidden)
        }
        .sheet(
            item: $store.scope(state: \.destination?.view5, action: \.destination.view5)
        ) { sheetStore in
            View5(store: sheetStore)
                .presentationDetents([.medium])
                .presentationDragIndicator(.hidden)
        }

**Comment by mbrandonw** [comment, marked as answer]
Hi @parvBhaskarStryker, there is no limitation set by TCA, but there may be limitations of the Swift compiler. Typically you can move the `sheet`'s to their own helper in the view to help out the compiler:

```swift
extension View {
  func destinations(store: Binding<StoreOf<Feature>>) -> some View {
    self
      .sheet(item: store.scope(…)) { … }
      .sheet(item: store.scope(…)) { … }
      .sheet(item: store.scope(…)) { … }
      .sheet(item: store.scope(…)) { … }
  }
}
```

And the do:

```swift
var body: some View {
  Form {
  }
  .destinations(store: $store)
}
```

That's the rough idea.

In the future these kinds of compilation problems will hopefully get better when we can delete the many overloads we have in the library. But we need to wait for 2.0 for that.

**Reply by tobitech** [reply]
Any idea how a binding is derived from an `@Perception.Bindable`. 
Here is how the store is defined `@Perception.Bindable var store: StoreOf<Feature>`
The `.destinations()` function requires a binding.
When I pass `$store` to the destinations function, I'm getting this error `Cannot convert value of type 'Bindable<StoreOf<Feature>>' (aka 'Bindable<Store<Feature.State, Feature.Action>>') to expected argument type 'Binding<StoreOf<Feature>>' (aka 'Binding<Store<Feature.State, Feature.Action>>')`

**Reply by mbrandonw** [reply]
Hi @tobitech, I didn't actually run my code, but did you try changing `Binding` to `Bindable` and see if that works?

**Reply by tobitech** [reply]
Hello @mbrandonw  Yes I did, just got more compiler errors `'Bindable' is ambiguous for type lookup in this context` and then `Cannot infer key path type from context; consider explicitly specifying a root type` in the scoping lines.

**Reply by mbrandonw** [reply]
If you are using the perception tools you will need `Perception.Bindable`.

**Reply by tobitech** [reply]
Oh yeah, I missed that. It compiles with `Perception.Bindable`. Thanks
